File: apps/ios/Sources/UI/Components/ComposeBar.swift
// Sublemonable — Copyright (C) 2026 Sublemonable contributors
// Licensed under the GNU Affero General Public License v3.0 or later.
// See the LICENSE file in the repository root for full license text.
// SPDX-License-Identifier: AGPL-3.0-only

import SwiftUI

/// Compose bar per design_system.components.compose_bar.
/// The send button is ALWAYS lemon yellow — it is the primary action colour.
public struct ComposeBar: View {
    @Binding var text: String
    @Binding var burnOnRead: Bool
    @Binding var ttlSeconds: Int?
    public var onSend: () -> Void

    @FocusState private var focused: Bool

    /// disappearing_messages.options_seconds from the master spec.
    public static let ttlOptions: [Int] = [30, 60, 300, 3600, 86400, 604800]

    public init(text: Binding<String>,
                burnOnRead: Binding<Bool>,
                ttlSeconds: Binding<Int?>,
                onSend: @escaping () -> Void) {
        _text = text
        _burnOnRead = burnOnRead
        _ttlSeconds = ttlSeconds
        self.onSend = onSend
    }

    public var body: some View {
        HStack(spacing: Spacing.s3) {
            ephemeralMenu

            TextField("", text: $text, axis: .vertical)
                .font(SubFont.body(TypeScale.message))
                .foregroundColor(.textPrimary)
                .tint(.lemon)
                .lineLimit(1...5)
                .focused($focused)
                .padding(.horizontal, Spacing.s4)
                .padding(.vertical, Spacing.s2 + 2)
                .background(
                    RoundedRectangle(cornerRadius: Radius.xl)
                        .fill(Color.backgroundElevated)
                )
                .overlay(
                    RoundedRectangle(cornerRadius: Radius.xl)
                        .strokeBorder(focused ? Color.borderActive : Color.borderToken,
                                      lineWidth: 1)
                )
                .overlay(alignment: .leading) {
                    if text.isEmpty {
                        Text("Encrypted message…")
                            .font(SubFont.body(TypeScale.message))
                            .foregroundColor(.textMuted)
                            .padding(.horizontal, Spacing.s4)
                            .allowsHitTesting(false)
                    }
                }

            sendButton
        }
        .padding(.horizontal, Spacing.s3)
        .padding(.vertical, Spacing.s2)
        .background(Color.backgroundSecondary)
        .overlay(alignment: .top) {
            Rectangle().fill(Color.borderToken).frame(height: 1)
        }
    }

    /// Attachment-position control: burn-on-read + TTL selection.
    private var ephemeralMenu: some View {
        Menu {
            Toggle(isOn: $burnOnRead) {
                Label("Burn on read", systemImage: "flame")
            }
            Menu("Disappear after") {
                Button("Off") { ttlSeconds = nil }
                ForEach(Self.ttlOptions, id: \.self) { option in
                    Button(Self.label(forTTL: option)) { ttlSeconds = option }
                }
            }
        } label: {
            Image(systemName: burnOnRead ? "flame.fill" : "timer")
                .font(.system(size: 18, weight: .medium))
                .foregroundColor(burnOnRead || ttlSeconds != nil ? .lemon : .textSecondary)
                .frame(width: 32, height: 32)
        }

    }

    private var sendButton: some View {
        Button(action: sendIfPossible) {
            ZStack {
                Circle().fill(Color.lemon)
                LemonSliceView(variant: .sendButton)
                    .frame(width: 22, height: 22)
            }
            .frame(width: 40, height: 40) // 40pt per spec
        }
        .buttonStyle(LemonSpringButtonStyle())
        .disabled(text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty)
        .opacity(text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? 0.45 : 1)
        .accessibilityLabel("Send encrypted message")
    }

    private func sendIfPossible() {
        guard !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return }
        onSend()
    }

    static func label(forTTL seconds: Int) -> String {
        switch seconds {
        case 30: return "30 seconds"
        case 60: return "1 minute"
        case 300: return "5 minutes"
        case 3600: return "1 hour"
        case 86400: return "1 day"
        case 604800: return "1 week"
        default: return "\(seconds) s"
        }
    }
}

/// Press: scale(0.92) with spring-back via easing_bounce, plus lemon glow.
public struct LemonSpringButtonStyle: ButtonStyle {
    public init() {}
    public func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .scaleEffect(configuration.isPressed ? 0.92 : 1.0)
            .shadow(color: Color.lemon.opacity(configuration.isPressed ? 0.4 : 0.0),
                    radius: 16)
            .animation(Motion.easeBounce(Motion.base), value: configuration.isPressed)
    }
}

File: apps/android/app/src/main/java/com/sublemonable/app/ui/components/ComposeBar.kt
// Sublemonable — Copyright (C) 2026 Sublemonable contributors
// Licensed under the GNU Affero General Public License v3.0 or later.
// See the LICENSE file in the repository root for full license text.
// SPDX-License-Identifier: AGPL-3.0-only

package com.sublemonable.app.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsFocusedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.LocalFireDepartment
import androidx.compose.material.icons.outlined.Timer
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.unit.dp
import com.sublemonable.app.ui.theme.BackgroundElevated
import com.sublemonable.app.ui.theme.BackgroundSecondary
import com.sublemonable.app.ui.theme.BorderActive
import com.sublemonable.app.ui.theme.BorderColor
import com.sublemonable.app.ui.theme.BurnOrange
import com.sublemonable.app.ui.theme.Lemon
import com.sublemonable.app.ui.theme.MonoFamily
import com.sublemonable.app.ui.theme.TextMuted
import com.sublemonable.app.ui.theme.TextPrimary
import com.sublemonable.app.ui.theme.TextSecondary
import com.sublemonable.app.ui.theme.TypeScale

/** Human label for a TTL option (features.messaging.disappearing_messages). */
fun ttlLabel(seconds: Int?): String = when (seconds) {
    null -> "Off"
    30 -> "30s"
    60 -> "1m"
    300 -> "5m"
    3600 -> "1h"
    86400 -> "1d"
    604800 -> "1w"
    else -> "${seconds}s"
}

/**
 * Compose bar (design_system.components.compose_bar):
 * #1A1800 surface with a 1px top border, pill input (#242100, lemon border
 * on focus), burn-on-read flame toggle, TTL cycler, and the 40dp lemon
 * circular send button (scale 0.92 on press, spring back).
 */
@Composable
fun ComposeBar(
    value: String,
    onValueChange: (String) -> Unit,
    onSend: () -> Unit,
    burnOnRead: Boolean,
    onToggleBurnOnRead: () -> Unit,
    ttlSeconds: Int?,
    onCycleTtl: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Column(modifier = modifier.fillMaxWidth()) {
        HorizontalDivider(thickness = 1.dp, color = BorderColor)
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .background(BackgroundSecondary)
                .padding(horizontal = 12.dp, vertical = 8.dp),
            verticalAlignment = Alignment.Bottom,
            horizontalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            // Burn-on-read toggle — flame lights orange when armed.
            IconButton(onClick = onToggleBurnOnRead) {
                Icon(
                    imageVector = Icons.Filled.LocalFireDepartment,
                    contentDescription = if (burnOnRead) {
                        "Burn on read enabled"
                    } else {
                        "Enable burn on read"
                    },
                    tint = if (burnOnRead) BurnOrange else TextSecondary,
                )
            }

            // TTL cycler — taps through Off/30s/1m/5m/1h/1d/1w.
            IconButton(onClick = onCycleTtl) {
                Column(horizontalAlignment = Alignment.CenterHorizontally) {
                    Icon(
                        imageVector = Icons.Outlined.Timer,
                        contentDescription = "Disappearing message timer: ${ttlLabel(ttlSeconds)}",
                        tint = if (ttlSeconds != null) Lemon else TextSecondary,
                    )
                    if (ttlSeconds != null) {
                        Text(
                            text = ttlLabel(ttlSeconds),
                            fontFamily = MonoFamily,
                            fontSize = TypeScale.Xs,
                            color = Lemon,
                        )
                    }
                }
            }

            // Pill input.
            val interactionSource = remember { MutableInteractionSource() }
            val focused by interactionSource.collectIsFocusedAsState()
            Box(
                modifier = Modifier
                    .weight(1f)
                    .heightIn(min = 40.dp)
                    .background(BackgroundElevated, RoundedCornerShape(24.dp))
                    .border(
                        width = 1.dp,
                        color = if (focused) BorderActive else BorderColor,
                        shape = RoundedCornerShape(24.dp),
                    )
                    .padding(horizontal = 16.dp, vertical = 10.dp),
                contentAlignment = Alignment.CenterStart,
            ) {
                BasicTextField(
                    value = value,
                    onValueChange = onValueChange,
                    modifier = Modifier.fillMaxWidth(),
                    textStyle = MaterialTheme.typography.bodyMedium.copy(color = TextPrimary),
                    cursorBrush = SolidColor(Lemon),
                    interactionSource = interactionSource,
                    maxLines = 5,
                    decorationBox = { innerTextField ->
                        if (value.isEmpty()) {
                            Text(
                                text = "Message",
                                style = MaterialTheme.typography.bodyMedium,
                                color = TextMuted,
                            )
                        }
                        innerTextField()
                    },
                )
            }

            // The send button is ALWAYS lemon yellow — primary action colour.
            LemonSendButton(
                onClick = onSend,
                enabled = value.isNotBlank(),
            )
        }
    }
}
